### ISSUE TITLE
Discussion for guidelines on removing HSTS preload rulesets

### ISSUE THREAD

**Issue by jeremyn** [EFForg/https-everywhere#7126]
@gloomy-ghost has submitted several pull requests deleting rulesets that are in the HSTS preload lists: [Chromium](https://cs.chromium.org/chromium/src/net/http/transport_security_state_static.json), [Firefox](https://dxr.mozilla.org/mozilla-aurora/source/security/manager/ssl/nsSTSPreloadList.inc), [Tor](https://gitweb.torproject.org/tor-browser.git/tree/security/manager/ssl/nsSTSPreloadList.inc?h=tor-browser-45.4.0esr-6.5-1). ( @gloomy-ghost has helpfully included links to these rulesets in their pull requests.) @J0WI and I have separately handled these and it turns out we disagree on when a ruleset should be deleted. We have mainly discussed this in pull request #7081 , but I'll repeat the discussion below.

@J0WI rejects rulesets for deletion if the `Strict-Transport-Security` header does not contain the `preload` header that is required in the [HSTS preload submission requirements](https://hstspreload.appspot.com/), even if the domain is in the preload lists. I assume that the `preload` header was in the lists when the domain was submitted to the lists and was later removed for whatever reason. This turns to be a common situation. My own opinion is that if the domain is in the preload lists and it has a `Strict-Transport-Security` header, then we can delete the ruleset even if `preload` is missing. @J0WI correctly points out that according to the submission requirements,

> Chrome has not yet removed any domains from the preload list for failing to keep up the requirements after submission, but there are plans to do so in the future.

I think it is too restrictive to not delete a ruleset in the preload lists just because it might be removed from the lists later.

Another problem that I found in pull request #7053 : https://wiki.openssl.org has a complete `Strict-Transport-Security:"max-age=31536000; includeSubDomains; preload"` header, but its redirect target https://wiki.openssl.org/index.php/Main_Page doesn't even have a `Strict-Transport-Security` header. Should we delete a ruleset if some of the responses have the header but some don't? If so, how many URLs do we have to check? What if only `preload` is missing?

Pinging @fuglede and @Hainish for their input.

**Comment by jeremyn** [comment]
@J0WI wrote [here](https://github.com/EFForg/https-everywhere/pull/7081#issuecomment-249434402):

> As @ScottHelme analyzed in his [blog post](https://scotthelme.co.uk/death-by-copy-paste/) some admins just copied the config from somewhere else and got added to the preload list without knowing the consequences.
So I think we should make sure, that they are in the current preload list (e.g. Tor), as well as in the preload list of future versions (e.g. mozilla central) and that they serve a valid HSTS preload header.

**Comment by jeremyn** [comment]
@J0WI You can't get added to the HSTS lists without having `preload` in the header. For the sites we're seeing that are in the lists but don't have `preload`, I assume the site admin had `preload` when they were added but then decided to remove it for some reason. So these sites aren't just blindly copy-and-pasting, in fact the opposite: they had `preload` at one point and then specifically decided to remove it from the header, possibly because they were already in the lists and didn't feel like they needed `preload` anymore.

**Comment by jeremyn** [comment]
Also, we should clearly document any requirements we have for the `Strict-Transport-Security` header. @gloomy-ghost has submitted [about 120 pull requests](https://github.com/EFForg/https-everywhere/pulls?utf8=%E2%9C%93&q=is%3Apr%20is%3Aopen%20author%3Agloomy-ghost%20rm) on this and it is not fair to them to let them do all that work if we are going to reject a lot of it because of undocumented requirements.

**Comment by J0WI** [comment]
I was not able to reproduce the issue in https://github.com/EFForg/https-everywhere/pull/7081#issuecomment-249433791, but since this already is a subdomain and the main host (`openssl.org`) is preloaded it's not required to send a full HSTS header there.

**Comment by J0WI** [comment]
I think we have to check the Alexa `top-*` rules more carefully if they really pass all HSTS requirements. If they are configured solid, it's likely to have them still preloaded in future. If they are not well configured, it's even more likely that Google will kick them off the list.

**Comment by J0WI** [comment]
Just a few more thoughts:
The motivation to remove HSTS preloaded rules is to shrink the extension and the resources it uses.
Before the preload list was shared between all major browsers, we had a `platform` for them. We could fetch all (pending) hosts from the preload to auto generate a huge ruleset (see #4396), but this would also blow up the extension.

**Comment by jeremyn** [comment]
For the https://wiki.openssl.org issue: on review I agree that HSTS works at the subdomain level, so if the header is good for https://openssl.org and includes subdomains, it is also good for https://wiki.openssl.org/index.php/Main_Page , which is what you said. So we agree. Also the [submission requirements](https://hstspreload.appspot.com/#submission-requirements) say:

> If you are serving an additional redirect from your HTTPS site, that redirect must still have the HSTS header (rather than the page it redirects to).

which is the behavior I'm seeing for the redirect I described, which is fine.

The header requirements should be the same for any ruleset, not just the Alexa ones, but I agree that as usual the higher Alexa domains should get more attention.

After more thought I've changed my mind and agree with you @J0WI about checking that the submission requirements are met before we delete a ruleset, to make sure that the site admin knows what they're doing. In practice this means:

* HTTPS works on base domain
* HTTP redirects to HTTPS on the base domain (without HTTPS Everywhere)
* The HTTPS page serves a valid `Strict-Transport-Security` header with `max-age`, `includeSubDomains`, and `preload`.

**Comment by jeremyn** [comment]
Also, if we reject a deletion pull request because the requirements aren't met, we should just close the pull request with a comment because that's not something the contributor can fix.

I've revert the deletions you noticed earlier, thanks for checking them.

**Comment by jeremyn** [comment]
@J0WI When we check the header, we should check the base domain and not worry about `www`. For example in https://github.com/EFForg/https-everywhere/pull/6995#issuecomment-249436976 , https://www.v2ex.com has a bad header, but https://v2ex.com is fine, so the deletion is valid. What do you think?

**Comment by J0WI** [comment]
I merged the revert commits and can confirm that v2ex.com sends a valid header.

As soon as we have a guideline about this, we can also add checks for it to our ruleset tests.

**Comment by jeremyn** [comment]
@J0WI Do you think we should stop reviewing and merging other HSTS deletion pull requests until we hear from the others? I think you and I agree on the rules.

**Comment by gloomy-ghost** [comment]
For `HTTP redirects to HTTPS on the base domain (without HTTPS Everywhere)`

What if the site does not support http anymore after the preload? Such as calomel.org

``` bash
$ curl -I https://calomel.org | grep preload
strict-transport-security: max-age=31556926; includesubdomains; preload
$ curl -I http://calomel.org --connect-timeout 30
curl: (28) Connection timed out after 30000 milliseconds
```

**Comment by gloomy-ghost** [comment]
I have created about 128 PRs if my record is correct. There are some statistical data:

HTTP Timeout: 1
Not sending header (anymore): 7
Incomplete certificate chain (Still working after cache intermediate) : 1

HSTS Header: 121
HSTS Preload: 105
HSTS Subdomain: 105
HSTS MaxAge > 18 weeks: 120

**Comment by J0WI** [comment]
> What if the site does not support http anymore after the preload? Such as calomel.org

That a requirement I not really agree with Google, but I think we should strictly follow their guideline.

**Comment by jeremyn** [comment]
For redirecting: once a site is HSTS preloaded, the HTTP version of the site becomes inaccessible, which I guess is why Google checks that HTTP redirects to HTTPS.

**Comment by jeremyn** [comment]
Something else to watch out for: for pull request #7077 , http://raymii.org returns a HTTP 200 but it uses a ["meta refresh"](https://en.wikipedia.org/wiki/Meta_refresh) to redirect to https://raymii.org/s/ . So unless this configuration was changed after it was accepted into the preload list, it means that:

* meta refresh redirects are valid, so a HTTP 200 does not mean the site is misconfigured, and
* the redirect can go to a different path on the same host, in this case `/` to `/s/`.

**Comment by jeremyn** [comment]
@J0WI You didn't answer the question in https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-249450970 but I'm assuming deletion pull requests are open for review now.

**Comment by jeremyn** [comment]
Another thing to look for: in pull request #7060 , http://whatwg.org redirects to http://www.whatwg.org , which redirects to https://whatwg.org . So unless the configuration was changed after it was accepted into the preload list, this means that:

* as long as the ultimate redirect destination of the HTTP URL is a HTTPS URL on the same host, it's okay if the redirect chain goes through other HTTP URLs or URLs with a different host, in this case `^` and `www`.

**Comment by jeremyn** [comment]
Another thing: according to [RFC 6797 section 6.1](https://tools.ietf.org/html/rfc6797#section-6.1):

> Directive names are case-insensitive.

So `includesubdomains` , all lower-case, is fine.

**Comment by gloomy-ghost** [comment]
Found an embarrassing instance: `eff.org` is preloaded but its header does not contain `preload`, and it does not provide http.

```
$ curl -I https://eff.org | grep Strict
Strict-Transport-Security: max-age=31536000; includeSubdomains
$ curl -I http://eff.org --connect-timeout 30
curl: (28) Connection timed out after 30001 milliseconds
```

<del>I don't know what to say</del>

**Comment by jeremyn** [comment]
Following my comment [here](https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-249597384) about redirecting, I think we should accept deletions for sites that have no HTTP. As I said, Google probably just cares that if you do have HTTP, it redirects to HTTPS, because once it's preloaded the HTTP site becomes inaccessible. But, it makes no sense for them to actually require an HTTP version of the site.

I'm not internal to the EFF and I don't know how their servers are configured, but making HTTP redirect to HTTPS is not completely trivial. It's possible to mess it up. The safest way to make sure you're not serving unencrypted traffic is not to respond to HTTP requests at all.

https://eff.org should have the `preload` directive though, no argument there.

**Comment by jeremyn** [comment]
https://www.translate.google.com is sort of funny. It's mismatched but is also in the preload list, so Firefox reports

>The owner of www.translate.google.com has configured their website improperly. To protect your information from being stolen, Firefox has not connected to this website.

>This site uses HTTP Strict Transport Security (HSTS) to specify that Firefox may only connect to it securely. As a result, it is not possible to add an exception for this certificate.

So Google has locked out one of their own sites due to HSTS preload.

**Comment by jeremyn** [comment]
Another thing: according to [RFC 6797 section 7.2](https://tools.ietf.org/html/rfc6797#section-7.2), it's technically a violation/misconfiguration for a HSTS compliant host to send the STS header over insecure HTTP. However we should still accept deletions for rules with this misconfiguration, because not only is this extremely minor, but it's possible that when we see the STS header over insecure HTTP, we're actually seeing a header injected by some malicious third party. There is no way to be sure over insecure HTTP.

**Comment by jeremyn** [comment]
Another thing: domains we remove should be in the released stable version of Firefox, Chromium, and Tor. We don't want to delete rulesets until the commonly-used preload lists have the equivalent behavior. For Firefox that means the domain should be in https://dxr.mozilla.org/mozilla-release/source/security/manager/ssl/nsSTSPreloadList.inc . I don't know what the equivalent source files are for Chromium or Tor.

**Comment by jeremyn** [comment]
@konklone In pull request https://github.com/EFForg/https-everywhere/pull/2681#r47963617 you wrote in code review:

> [uspsoig.gov is HSTS preloaded](https://github.com/EFForg/https-everywhere/pull/@USGPO) with an includeSubDomains policy, so you can use `*.uspsoig.gov` here instead of just `www`.

According to my understanding of HSTS and the discussion in this issue, `includeSubDomains` means exactly that `*.uspsoig.gov` is effectively in the preload list along with `uspsoig.gov`, so for example https://www.uspsoig.gov and https://kce.uspsoig.gov are preloaded. So why did you advise replacing `www.uspsoig.gov` with `*.uspsoig.gov`?

**Comment by J0WI** [comment]
What about detected errors? I'm not sure how Mozilla handles them during the release process.
E.g. in case of #7027: https://dxr.mozilla.org/mozilla-central/rev/9baec74b3db1bf005c66ae2f50bafbdb02c3be38/security/manager/ssl/nsSTSPreloadList.errors#1573

**Comment by jeremyn** [comment]
I'm not sure what you are asking.

**Comment by gloomy-ghost** [comment]
There will be a false value to disable it. 
https://dxr.mozilla.org/mozilla-release/source/security/manager/ssl/nsSTSPreloadList.inc#318

**Comment by gloomy-ghost** [comment]
Well, [konklone.com](https://dxr.mozilla.org/mozilla-release/source/security/manager/ssl/nsSTSPreloadList.inc#4829) is still in the release list, but you can't find it in aurora anymore.

**Comment by J0WI** [comment]
I just asked the Mozillian that implemented this Feature in Firefox and there is now a wiki page about the preload list:
https://wiki.mozilla.org/SecurityEngineering/HTTP_Strict_Transport_Security_(HSTS)_Preload_List

**Comment by jeremyn** [comment]
Can anyone figure out how it is that `googlesource.com` isn't in https://dxr.mozilla.org/mozilla-release/source/security/manager/ssl/nsSTSPreloadList.inc but for example http://gerrit.googlesource.com seems preloaded in Windows 10, Firefox 49.0.1?

**Comment by J0WI** [comment]
It is on beta. A new stable release will likely ship this next week.

**Comment by jeremyn** [comment]
@J0WI `beta` is a different channel though. Am I somehow running `beta`, or do they push the `beta` list to stable users more frequently than `release`? The [Mozilla wiki](https://wiki.mozilla.org/SecurityEngineering/HTTP_Strict_Transport_Security_(HSTS)_Preload_List) link you gave earlier is great, thanks, but I don't think it answers this question.

**Comment by J0WI** [comment]
As far as I understand the article only central and aurora are updated periodically. The list in beta and release follow the normal release cycle.

**Comment by jeremyn** [comment]
I just noticed that the `mozilla-release` URL in the wiki article, https://hg.mozilla.org/releases/mozilla-release/file/tip/security/manager/ssl/nsSTSPreloadList.inc , has a latest commit date of Oct 6 2016. Since I updated my Firefox browser to 49.0.1 before Oct 6 2016, then `mozilla-release` doesn't track what is currently in my browser like I thought it did. Basically I thought `mozilla-release` is identical to what's in my stable release browser, but that must not be true.

**Comment by konklone** [comment]
> According to my understanding of HSTS and the discussion in this issue, includeSubDomains means exactly that *.uspsoig.gov is effectively in the preload list along with uspsoig.gov, so for example https://www.uspsoig.gov and https://kce.uspsoig.gov are preloaded. So why did you advise replacing www.uspsoig.gov with *.uspsoig.gov?

@jeremyn Because at the start of that thread, I thought that HTTPS Everywhere wanted to use the preload list as a source of rules to **add**, since the domains had expressed a desire to be hardcoded as HTTPS. If you look later [in that thread](https://github.com/EFForg/https-everywhere/pull/2681#r47963617), @jsha mentions that this isn't the case, and so I suggest removing `uspsoig.gov`, along with others.

**Comment by konklone** [comment]
> > Chrome has not yet removed any domains from the preload list for failing to keep up the requirements after submission, but there are plans to do so in the future.
> 
> I think it is too restrictive to not delete a ruleset in the preload lists just because it might be removed from the lists later.

My understanding is that Mozilla already removes entries in the list that respond over HTTPS but which drop the `preload` directive. Chrome has suggested they will do this in the future, in part to make the removal process simpler and more automatable (if you want to be removed, just drop the `preload` directive!). Other browsers/OSes that use the preload list might do something similar.

> @J0WI You can't get added to the HSTS lists without having preload in the header. 

It is possible to be added to the HSTS preload list manually, without having `preload` in the header. You can see a "manual" section later on in the file starting with `// START OF MANUAL ADDITIONS`. (`login.gov` is one of them, though it serves a `preload` directive anyway). I believe these manual entries would _not_ be auto-pruned by Chrome, even if they started auto-pruning based on the presence of a `preload` directive.

cc @lgarron

**Comment by lgarron** [comment]
> I believe these manual entries would not be auto-pruned by Chrome, even if they started auto-pruning based on the presence of a preload directive.

Indeed. Definitely not without thorough discussion.

> cc @lgarron

I'm here!

**Comment by jeremyn** [comment]
@konklone Thanks for clarifying your earlier comments.

Later in this thread (https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-249436544) I changed my mind and agreed with @J0WI that we shouldn't delete rulesets for preloaded domains unless the domains still meet the preload conditions.

My reading of the Mozilla wiki link in https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-250908333 is that Mozilla starts with the Chromium list and does some minimal checks before updating its own list. Particularly,

> (the "preload" directive is ignored)

which I think means that as long as the domain is in the Chromium list, the STS header can leave out `preload` and Mozilla will still include the domain in its own list.

**Comment by jsha** [comment]
Hi all! Just getting back from vacation, sorry for the delay here.

The reasoning behind deleting / not including hosts that are already in the browser HSTS preload lists is exactly as @J0WI said: It's redundant to include them in the extension since they are already in the browser, and we can save a little extension size by removing them. There's also a small side benefit: Since we don't automatically pull from browser preload lists, rulesets that include copies of those hostnames may go stale and break.

That said, the number of hostnames in HTTPS Everywhere's rulesets is much higher than in any of the HSTS preload lists, so the size impact of over-including is relatively small. Therefore I think it makes sense to err on the side of including. So, e.g. HTTPS Everwhere could include a hostname if:

 - It's in the preload list for one but not all supported browsers.
 - It's removed from the preload list in an upcoming release of any supported browser.
 - It's likely to be removed from the preload list in an upcoming release of any supported browser, because the host no longer meets program requirements for that browser.

What do you all think of this approach?

**Comment by lgarron** [comment]
> That said, the number of hostnames in HTTPS Everywhere's rulesets is much higher than in any of the HSTS preload lists, so the size impact of over-including is relatively small. Therefore I think it makes sense to err on the side of including.

I think this is a good, simple, correct answer.

Note that:
1) HTTPS Everywhere is still useful even after baked-in browser values expire (cf. recent Firefox/Tor issue with expired pins for a similar scenario), and
2) browser policies are not inherently subject to any promises – any of them could theoretically change at a moment's notice, or even without notice.

**Comment by jeremyn** [comment]
It's frustrating that this discussion so far has been almost entirely technical with no consideration of human, likely unpaid volunteer time as one of the limited resources available. In fact there are several open discussions such as https://github.com/EFForg/https-everywhere/issues/6863 , https://github.com/EFForg/https-everywhere/issues/6912 , https://github.com/EFForg/https-everywhere/issues/7242 , https://github.com/EFForg/https-everywhere/issues/7256 , https://github.com/EFForg/https-everywhere/issues/7258 (related to HSTS), https://github.com/EFForg/https-everywhere/issues/7277 , and PR https://github.com/EFForg/https-everywhere/pull/7255 that one way or another are about automation/saving time

**Comment by jeremyn** [comment]
@gloomy-ghost has kindly submitted about 150 HSTS removal pull requests and @J0WI and I have merged most of them. By now my review process is essentially scriptable. For reference, it's:

For each `$DOMAIN` (ex: `example.com`, `example.net`) in the ruleset:

* Check that `$DOMAIN` is in the [Chromium](https://chromium.googlesource.com/chromium/src.git/+/53.0.2785.143/net/http/transport_security_state_static.json), [Firefox](https://dxr.mozilla.org/mozilla-release/source/security/manager/ssl/nsSTSPreloadList.inc), and [Tor](
https://gitweb.torproject.org/tor-browser.git/tree/security/manager/ssl/nsSTSPreloadList.inc?h=tor-browser-45.4.0esr-6.5-1) lists and that it includes subdomains. (@gloomy-ghost has been nice enough to link to the exact line for their rulesets.)
* `curl -I http://$DOMAIN` , confirm that it redirects to `https://$DOMAIN`
* `curl -I https://$DOMAIN` , confirm that it returns some valid HTTP code and that it has `Strict-Transport-Security: max-age=$LONG_ENOUGH ; includeSubDomains ; preload`
* `curl -I https://www.$DOMAIN` , confirm that it either returns some valid HTTP code or `www.$DOMAIN` is not in DNS (it's almost always the former)
* Disable HTTPS Everywhere in stable Firefox and check that `http://$DOMAIN` is rewritten to HTTPS in dev tools
* Disable HTTPS Everywhere in stable Chrome and check that `http://$DOMAIN` is rewritten to HTTPS in dev tools, shown as a request to `http` returning `HTTP 307` HSTS internal redirect
* If the deleted ruleset refers to other rulesets, for example it's one of the many child Microsoft or Google sub-rulesets, then delete any references to the deleted ruleset from the other ruleset(s).

If these items pass, I merge the removal. My earlier comments in this issue describe handling edge cases, such as directive casing and redirect chains.

Other things that I don't necessarily do but wouldn't hurt:
* Search the repository for `$DOMAIN` for references not listed in the deletion candidate
* Check HSTS preloading in Tor

**Comment by jeremyn** [comment]
My previous comment was technical, this comment is procedural.

If someone submits a removal pull request and it passes the test, I remove it. If someone submits a new ruleset, and I happen to check if it's preloaded, and it is, I don't add it.

I don't agree with keeping redundant HTTPS Everywhere coverage for fully preloaded domains. Each ruleset is a potential maintenance hassle for multiple people: the site owners, unaffiliated contributors, and reviewers. If someone goes to the trouble of preloading their site, then removing that domain from HTTPS Everywhere saves human maintenance time, even if the clock/RAM/etc resources used by having that extra ruleset are minimal.

It's frustrating that this discussion so far has been almost entirely technical with no consideration of human, likely unpaid volunteer time as one of the limited resources available. In fact there are several open discussions such as https://github.com/EFForg/https-everywhere/issues/6863 , https://github.com/EFForg/https-everywhere/issues/6912 , https://github.com/EFForg/https-everywhere/issues/7242 , https://github.com/EFForg/https-everywhere/issues/7256 , https://github.com/EFForg/https-everywhere/issues/7258 (related to HSTS), https://github.com/EFForg/https-everywhere/issues/7277 , and PR https://github.com/EFForg/https-everywhere/pull/7255 that one way or another are about automation/saving time , and most of them are just echo chambers for volunteers. I don't want to get into my own personal concerns in the middle of this random issue, but I think there is a problem and the discussion in this issue is related to it.

**Comment by galeksandrp** [comment]
@jeremyn `curl "https://hstspreload.appspot.com/api/v2/preloadable?domain=$DOMAIN"` will show you errors and warnings, and `curl "https://hstspreload.appspot.com/api/v2/status?domain=$DOMAIN"` show current status.

**Comment by jeremyn** [comment]
@galeksandrp That's a good tip, thanks. I tried to find documentation for that at the appspot link but couldn't. I found this though https://scotthelme.co.uk/hsts-preloading/ which also mentions `https://hstspreload.appspot.com/api/v2/preloadable?domain=$DOMAIN` . I'll try these along with my normal checking for a little while to see how they match up.

**Comment by lgarron** [comment]
> @galeksandrp That's a good tip, thanks. I tried to find documentation for that at the appspot link but couldn't. I found this though https://scotthelme.co.uk/hsts-preloading/ which also mentions https://hstspreload.appspot.com/api/v2/preloadable?domain=$DOMAIN . I'll try these along with my normal checking for a little while to see how they match up.

Indeed, that endpoint is not prominently documented, but it is fairly stable.
If you get back a `200 OK` response, the `errors` and `warnings` fields in the JSON response should be reliable.

**Comment by lgarron** [comment]
> Indeed, that endpoint is not prominently documented, but it is fairly stable.
> If you get back a 200 OK response, the errors and warnings fields in the JSON response should be reliable.

Also note that requirements for removing/pruning entries may not be the exact complement of "preloadable".
So you can't quite assume that something will be **removed** if it's no longer preloadable.
But if it's preloaded and still matches preload requirements, it will probably stay. I'm working on a more robust mechanism to scan and surface such a status, though. (Follow https://github.com/chromium/hstspreload.appspot.com/issues/35 for updates.)

**Comment by jsha** [comment]
@jeremyn you make a good point. I definitely agree that the time and effort of ruleset maintainers are vital resources, that we should use as efficiently as possible. That's why I originally added the automated ruleset tests. Hopefully we can continue to improve them!

> I don't agree with keeping redundant HTTPS Everywhere coverage for fully preloaded domains.

I think this is the current consensus on this issue. Was someone stating differently or did I miss something? I think there are some edge cases, like the ones I mentioned above. Do you think we should err on the side of inclusion or exclusion for those edge cases?

> In fact there are several open discussions such as #6863 , #6912 , #7242 , #7256 , #7258 (related to HSTS), #7277 , and PR #7255 that one way or another are about automation/saving time , and most of them are just echo chambers for volunteers.

These are good issues, and I'm sorry we haven't seen more action / participation on them. @Hainish and @cooperq are doing most of the management work for our extensions these days. I know @Hainish is getting back from a vacation tomorrow and hopefully will be able to do a run through the issues.

**Comment by lgarron** [comment]
> Was someone stating differently or did I miss something? I think there are some edge cases, like the ones I mentioned above. Do you think we should err on the side of inclusion or exclusion for those edge cases?

Yep, me. I think it's worthwhile for HTTPS Everywhere to aim to be a complete list without any assumptions or dependencies on browser behaviour. But that's an abstract preference, not rooted in bloat/maintainability constraints.

**Comment by fuglede** [comment]
Similarly, there's https://hstspreload.com/api/v1/status/example.com which makes it fairly straightforward to check coverage in other browsers. I've used that one myself for a couple of PRs to make sure that all relevant ones are covered.

**Comment by jeremyn** [comment]
I want to err on the side of not deleting rulesets. Without firm guidance from the Chromium and Firefox teams on a removal checklist, we have to use vague terms like @lgarron uses in https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-252787320 : "may not", "can't quite assume", "probably". So my approach is that if a ruleset wouldn't be preloaded today, then we don't delete the ruleset, even if it's already preloaded.

Also as @J0WI pointed out (https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-249434713), HSTS is potentially both confusing and dangerous for site owners. So checking that the site meets all preload conditions also helps confirm that the site owner knows what they're doing.

@lgarron I understand the elegance in wanting the add-on to be independent of what particular browsers do. We're introducing a quiet assumption that HTTPS Everywhere will only be used on browsers with the HSTS preload list. But it does impose a maintenance burden. Also I'm personally okay with thinking of HTTPS Everywhere as augmenting specific browsers rather than as completely standalone software.

**Comment by lgarron** [comment]
> But it does impose a maintenance burden.

Hmm, what if you included all preloaded entries automatically?
That would cause some bloat, but would be easy to automate.

**Comment by jeremyn** [comment]
@lgarron Extending my previous comment, consider that by deleting HSTS preloaded rulesets, we're accepting a dependency on browser HSTS preloading. This is like adding a dependency on an external library in a more traditional software product.

You're proposing that we copy the library into our software to remove the dependency by internalizing it. So what are the advantages of disadvantages of doing that? We've discussed disadvantages: increased programmer maintenance, slightly increased computing resources. But, the only advantage I can think of is a theoretical distrust of dependencies. I can't think of any practical advantages, meaning any problems we actually have at the moment that internalizing the dependency would solve.

**Comment by lgarron** [comment]
> I can't think of any practical advantages, meaning any problems we actually have at the moment that internalizing the dependency would solve.

This bug exists because there is regular discussion about whether to include certain entries from the preload list, and how to pick them. Including all values is at least as safe as any intermediate option, and is easy to do objectively.

> We're introducing a quiet assumption that HTTPS Everywhere will only be used on browsers with the HSTS preload list.

Indeed, but that's not the only way it's being used.

For example:
- It seems [Brave uses HTTPS Everywhere without the HSTS preload list](https://twitter.com/xxdesmus/status/701185424153141248).
- Cloudflare [announced](https://blog.cloudflare.com/fixing-the-mixed-content-problem-with-automatic-https-rewrites/) that they are using HTTPS Everywhere to power their "Automatic Rewrites". Now, in this case they are combining  it with Chrome's HSTS preload list, but maybe some other service will introduce a similar feature and just use HTTPS Everywhere – each time the authors would need to make a decision to combine them and write the infrastructure to support both.

**Comment by jeremyn** [comment]
@lgarron It should be easy for Brave or anyone to programmatically generate a huge ruleset that mimics the HSTS preload list. In fact it would be helpful if they did that and contributed the process upstream to us, particularly a process that makes nice rulesets given a list of domains for us to use for e.g. https://github.com/EFForg/https-everywhere/issues/6912 .

The two extreme options are to fully accept HSTS duplicate rulesets ("Including all entries") on one hand, and to both reject new duplicates and programmatically search for and purge existing duplicates on the other. A compromise would be to process individual deletion PRs but otherwise leave existing rulesets alone, and make some effort not to accept new duplicates. That's what we're doing at the moment.

I don't like talking about what's easy in the abstract because easy only makes sense for a given group of people, over a given timeframe, with given constraints. For example it's easiest for everyone here if we get rid of the "ignore niche browsers like Brave" constraint entirely. (FWIW I actually do think we should ignore them, or at least make them pay for the cost of their own customizations.) It's easiest for contributor @gloomy-ghost if we just ask them to stop submitting deletion PRs. It's easiest for me as a reviewer to blindly accept any open PRs @gloomy-ghost has already submitted. Etc etc.

I'm sorry if this seems pedantic, but we just have multiple players with conflicting costs and viewpoints (see my earlier mini-rant https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-252771916), and with constrants/goals that aren't totally clear, so it's important to clarify these things.

**Comment by galeksandrp** [comment]
```sh
echo '<ruleset name="HSTS">' > HSTS.xml
wget https://chromium.googlesource.com/chromium/src.git/+archive/master/net/http.tar.gz -O - | tar xz transport_security_state_static.json -O | sed 's&^ *//.*&&' | jq -r '.entries[].name' | xargs -n1 -i echo -e '\t<target host="{}" />' >> HSTS.xml
echo '
	<rule from="^http:" to="https:" />
</ruleset>' >> HSTS.xml
```

**Comment by jeremyn** [comment]
@galeksandrp Pretty much, nice job.

**Comment by lgarron** [comment]
> and to both reject new duplicates and programmatically search for and purge existing duplicates on the other.

**Comment by lgarron** [comment]
> to both reject new duplicates and programmatically search for and purge existing duplicates on the other.

**Comment by lgarron** [comment]
Sorry if I wasn't clear in my latest post; I'm no longer talking about the abstract.

> to both reject new duplicates and programmatically search for and purge existing duplicates on the other.

From what I understand, a safe and completely automatable is to do this, combined with loading the preload list using something like @galeksandrp's script.
What are your main concerns with that?

**Comment by galeksandrp** [comment]
`TEST=rules test/travis.sh`

**Comment by Hainish** [comment]
Just catching up on this thread after returning from vacation.

It seems most reasonable to me that we only delete preloaded rulesets we expect will not be removed (e.g. due to the absence of the `preload` directive) and that the [checks](https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-252771754) @jeremyn is doing to the PRs @gloomy-ghost is submitting is in good due diligence towards this objective.  I generally agree that we should exclude the list of preloaded domains from the rulesets and not reproduce them in a separate rule, due to the fact that rules are loaded into memory and thus increase the memory footprint of the extension as increasing download size.

For downstream dependencies such as Brave and Cloudflare, I agree that the burden of correct implementation of the rulesets is on them, and that we may want to automate the process of creating an HSTS preload ruleset which is not included in the extension itself for these dependencies, but this should be considered a convenience and not an obligation of HTTPS Everywhere.  Seems like what might be more reasonable than creating a script to generate a ruleset which we include in our ruleset path is just making such a script available for these dependencies which they can run themselves as a build step.

**Comment by galeksandrp** [comment]
```js
if (navigator.userAgent.indexOf('Brave') !== -1) {
all_rules.addFromXml(loadExtensionFile('rules/hsts.rulesets', 'xml'));
}
```

Here are list of not yet removed domains https://travis-ci.org/galeksandrp/https-everywhere/builds/166926870

Also, why Chromium preload list contains 3-rd level domain names
```json
    { "name": "api.simple.com", "mode": "force-https" },
```
when https://hstspreload.appspot.com requires `includeSubdomains` for adding?

**Comment by gloomy-ghost** [comment]
Some domains are only preloaded in Chrome.

e.g, www.gov.uk

**Comment by galeksandrp** [comment]
```sh
wget https://chromium.googlesource.com/chromium/src.git/+archive/master/net/http.tar.gz -O - | tar xz transport_security_state_static.json -O | sed 's&^ *//.*&&' | jq -r '.entries[].name' | sort > chromium.txt
curl https://hg.mozilla.org/mozilla-central/raw-file/7ae377917236b7e6111146aa9fb4c073c0efc7f4/security/manager/ssl/nsSTSPreloadList.inc | grep '  /\* "' | sed 's&^  /\* "&&' | cut -d '"' -f 1 > mozilla.txt
comm -23 chromium.txt mozilla.txt
```
3653 domains, wow. But not www.gov.uk, it is preloaded in Chromium master.

**Comment by jeremyn** [comment]
@lgarron I don't want to get too derailed by word arguments, but it's not clear what you mean by "easy": easy for whom? Anyway.

It would be great if someone came up with a process that purged HSTS preloaded domains. I certainly don't want @gloomy-ghost to feel they should submit thousands (?) of PRs to remove these rulesets. But, we would have to review this process, document it, discuss how to use it etc.

The code-golfed one liner @galeksandrp provided is a good proof-of-concept and produces output vaguely like what I had in mind, but it's not something we should include as-is. It is not easily maintainable for several reasons.

I agree with @Hainish that we shouldn't include a HSTS-mega-ruleset.

**Comment by jeremyn** [comment]
@galeksandrp 

>in `mozilla-central` (is it master?). 

I don't know how the branches relate to the Firefox I'm actually running, see https://github.com/EFForg/https-everywhere/issues/7126#issuecomment-252275086 . It would be nice if Mozilla could clarify. This is partly why I manually check HSTS preloading in Firefox even if it's in the `mozilla-release` list.

**Comment by galeksandrp** [comment]
#5851

**Comment by jeremyn** [comment]
@RReverser I notice you just added https://github.com/EFForg/https-everywhere/commit/c32bc6c9d500341a2fcd4de881c1a086e14ddf4f "Add script for HSTS conversion." Could you please describe how that fits into the discussion in this issue?

**Comment by RReverser** [comment]
@jeremyn I don't know if it's related to this discussion (didn't have time to read it yet). @Hainish just asked to add our script there if we can share it, so I did in hope it will be helpful 😄

**Comment by Hainish** [comment]
This is relevant to the discussion here as the script is the method by which Cloudflare generates their HSTS ruleset.  This can be provided as a convenience to downstream dependencies (e.g. Brave or others).

**Comment by mnordhoff** [comment]
@jeremyn I don't know Mozilla's full release process, but i know the basic bits. `mozilla-release` is the release repository. They make new commits as-needed that will be in the next point release. (Including automated HPKP and HSTS preload list updates.) So the repository tip will usually be somewhat newer than whatever the last release was. The releases are tagged, e.g. [`FIREFOX_49_0_1_RELEASE`](https://hg.mozilla.org/releases/mozilla-release/rev/FIREFOX_49_0_1_RELEASE). I don't know of a floating "current release" tag or other indicator.

There's also the matter of ESR releases, which are in separate version-specific repositories, e.g. https://hg.mozilla.org/releases/mozilla-esr38/ (EOL) and https://hg.mozilla.org/releases/mozilla-esr45/ (current), with their own similar release tags, e.g. [`FIREFOX_45_4_0esr_RELEASE`](https://hg.mozilla.org/releases/mozilla-esr45/rev/FIREFOX_45_4_0esr_RELEASE).

There may be one, but i don't know of a fully-automated way to download the current preload lists, or at least a simple way. You could parse https://hg.mozilla.org/releases/ for the most recent ESR repository, parse `hg tags` or https://hg.mozilla.org/releases/mozilla-release/tags (plus the ESR repository page) for the most recent Firefox release, &c, but it's kind of hacky and unreliable.

I would also want to wait a few days between a new Firefox release and removing things from HTTPS Everywhere, to give users time to upgrade. (Though that would only be a problem on the rare occasions the respective release schedules overlapped.) There is also overlap between the ESR series ([38](https://github.com/mnordhoff/firefox-pin-data/blob/af96c810f5c191713b0192b69af044785d098ae6/firefox-esr38-releases.csv#L16), [45](https://github.com/mnordhoff/firefox-pin-data/blob/af96c810f5c191713b0192b69af044785d098ae6/firefox-esr45-releases.csv#L2), so it would be most diligent to check *both* series as long as the older one is still supported.

\o/

**Comment by jeremyn** [comment]
Thanks @mnordhoff . My interest is knowing which tag/HSTS file my current browser is using. It looks like that is the `RELEASE` tags like [`FIREFOX_49_0_1_RELEASE`](https://hg.mozilla.org/releases/mozilla-release/rev/FIREFOX_49_0_1_RELEASE) which you linked. I don't know if we want to check ESR, other than the version included with Tor, and if so which versions. I haven't been checking them.

**Comment by Hainish** [comment]
@mnordhoff the most reliable way to determine the current release of ESR and stable that I've found is:

```shell
STABLE_VERSION=$(curl -D /dev/stdout "https://download.mozilla.org/?product=firefox-latest&os=linux64&lang=en_US" 2>&1 | sed -n '/Location: /{s|.*/firefox-\(.*\)\.tar.*|\1|p;q;}') && \
DEV_VERSION=$(curl -D /dev/stdout "https://download.mozilla.org/?product=firefox-aurora-latest-ssl&os=linux64&lang=en-US" 2>&1 | sed -n '/Location: /{s|.*/firefox-\(.*\)\.tar.*|\1|p;q;}') && \
ESR_VERSION=$(curl -D /dev/stdout "https://download.mozilla.org/?product=firefox-esr-latest&os=linux64&lang=en_US" 2>&1 | sed -n '/Location: /{s|.*/firefox-\(.*\)\.tar.*|\1|p;q;}')
```

This is the method that we use in the https://github.com/EFForg/https-everywhere-docker-base/ repo.

Once we get that, I think it makes sense to checkout the tag corresponding to whichever release you're interested in, and then looking at the preload list included there.

**Comment by mnordhoff** [comment]
Sounds legit. Or sort of legit. With a little more `sed` and `curl` it's enough to download the current lists from hgweb, or check them out locally.

**Comment by Hainish** [comment]
A little guidance here:

We should only remove target domains which are HSTS preloaded if the preload is active for all currently supported versions of Firefox (ESR, Stable, and Dev) as well as the current version of Chromium.

I'll add this to the `CONTRIBUTING.md`, and I've created a new task to script this process: https://github.com/EFForg/https-everywhere/issues/8122

**Comment by Hainish** [comment]
https://github.com/EFForg/https-everywhere/pull/8127 is a node utility which fetches the HSTS rulesets from Firefox {Dev, Stable, ESR} and Chromium (head), and deletes targets contained in all of them.  It also deletes rulesets completely if all targets would have otherwise been deleted.

Once this is merged we can incorporate this into the release process.

**Comment by Hainish** [comment]
https://github.com/EFForg/https-everywhere/pull/8128 is the result of running the abovementioned utility

**Comment by Hainish** [comment]
https://github.com/freedomofpress/securedrop/issues/1517#issuecomment-271698153 provides additional context for the process of HSTS removal, which has not been formalized just yet.  I suspect once this process is formalized, we will have to revisit the HSTS pruning utility in https://github.com/EFForg/https-everywhere/pull/8127

**Comment by lgarron** [comment]
> freedomofpress/securedrop#1517 (comment) provides additional context for the process of HSTS removal, which has not been formalized just yet.

To add more context, the first step will likely be to start removing "new" (since Feb. 29, 2016) entries that respond over HTTPS, but do not send a header that does not have one of:
- `max-age` >= 10886400
- `includeSubDomains`
- `preload`

Anything other removal conditions will be tricky, but we will need to look at them carefully.

**Comment by Hainish** [comment]
Thanks, @lgarron - I'll add these additional conditions to our internal hsts pruning utility.

**Comment by cooperq** [comment]
@Hainish lmk when this is ready for a final review

**Comment by Hainish** [comment]
@lgarron for our hsts pruning utility, we attempt to exempt HTTPS Everywhere target domains from removal if they are likely to be removed from the preload list, to ensure that once you remove these domains they will not be left unprotected by HTTPS Everywhere as well.  This means our exemption logic should match up with your removal logic.  To this end, I have a few questions:

1. Will any of the above directives be case sensitive?
2. Are you going to be checking for automatic redirection from HTTP to HTTPS?

**Comment by lgarron** [comment]
> Will any of the above directives be case sensitive?

No, since the spec is not case-sensitive.

> Are you going to be checking for automatic redirection from HTTP to HTTPS?

The scans will be checking for it, but I don't know if it's safe to include in the policy yet.
The initial policy is meant to avoid allowing an attacker to drop the site from the preload list unless they can respond with a valid cert from the site, so we probably won't start with it – but in principle I'd like to keep enforcing all requirements.

I recommend the more conservative approach, but talk to me if the numbers look burdensome?

**Comment by Hainish** [comment]
@lgarron in our case, the most conservative approach is to use the most stringent conditions in order to limit the amount of rules removed.  We'd like to avoid as much as possible removing target domains that may be removed from the preload list in the future.  For this reason we'll implement checking for HTTP redirects as well, in case you implement that in the future.

**Comment by lgarron** [comment]
> This means our exemption logic should match up with your removal logic.

Also, if you want to test against the actual requirements checker, https://github.com/chromium/hstspreload should be easy to run if you have `go`.

**Comment by Hainish** [comment]
To wrap up this discussion, the guidelines on removing HSTS preload targets are formalized in the [`hsts-prune` utility](https://github.com/EFForg/https-everywhere/pull/8127), but to be explicit:

> Let `included domain` denote either a `target`, or a parent of the `target`.  Let `supported browsers` include the ESR, Dev, and Stable releases of Firefox, and the Stable release of Chromium.  If `included domain` is a parent of the `target`, the `included domain` must be present in the HSTS preload list for all `supported browsers` with the relevant flag which denotes inclusion of subdomains set to `true`.  If `included domain` is the `target` itself, it must be included the HSTS preload list for all `supported browsers`.  Additionally, if the http endpoint of the `target` exists, it must issue a 3XX redirect to the https endpoint for that target.  Additionally, the https endpoint for the `target` must deliver a `Strict-Transport-Security` header with the following directives present:
> 
> - `max-age` >= 10886400
> - `includeSubDomains`
> - `preload`
> 
> If all the above conditions are met, a contributor may remove the `target` from the HTTPS Everywhere rulesets.  If all targets are removed for a ruleset, the contributor is advised to remove the ruleset file itself.  The ruleset `rule` and `test` tags may need to be modified in order to pass the ruleset coverage test.

**Comment by Hainish** [comment]
I'll include this in the documentation for `CONTRIBUTING.md` that I'm working through
